firedrake/bcs.py

```diff
@@ -159,6 +160,13 @@ def set_value(self, val):
         self._original_val = val
 
     @utils.cached_property
+    def domain_args(self):
+        """The sub_domain the BC applies to."""
+        if isinstance(self.sub_domain, str):
+            return (self.sub_domain, )
+        return (as_tuple(self.sub_domain), )
```

Comment:
Arguably this is wrong, but it works for the use-cases we need.

Additionally, we should also provide a copy constructor for DirichletBC.